dpti/workflow.py

```
@@ -0,0 +1,385 @@
+    name = step["name"]
+    if "command" not in step:
+        raise ValueError(f"workflow step '{name}' does not define a command")
+    work_dir = _step_work_dir(root, step)
+    command = _normalize_command(step["command"])
+    print(f"[workflow] run step: {name}")
+    print(f"[workflow] cwd: {work_dir}")
+    print("[workflow] command:", " ".join(shlex.quote(item) for item in command))
+    if dry_run:
+        return {"returncode": 0}
+    work_dir.mkdir(parents=True, exist_ok=True)
+    completed = subprocess.run(command, cwd=str(work_dir))
```

Ruff (0.15.15)

[error] `subprocess` call: check for execution of untrusted input

(S603)